### PR TITLE
add middlewares to module system, codegen and inject clients

### PR DESCRIPTION
**Pull request by jakelarkn**
@uber/zanzibar-team 

key changes
- add middleware as a class type in the module system
- needed config changes:  each middleware has its own middleware-config.json which conforms to the module system json shape
- codegen now does the following
  - generates the module dependency struct for each middleware
  - generates a middleware factory instance which 1) holds a ref to the dependencies struct  2) exposes a factory method for MiddlewareHandle, for the endpoints to pass in their respective options

### CODE REVIEW

File: test/lib/compare.go [hunk 1]
```diff
@@ -68,3 +79,20 @@ func DiffStrings(t *testing.T, exp, actual string) {
 	}
 	t.Errorf("Unexpected result: \n%s\n", d)
 }
+
+func deepEqualJSON(b1 []byte, b2 []byte) (bool, error) {
```
Comment (cl1337) [inline test/lib/compare.go]: mm ....  how about
```
	a := `{"a":{"b":1}, "c":{"a":2}}`
	b := `{"c":{"a":2}, "a":{"b":1}}`

	assert.JSONEq(t, a, b)
```

File: backend/repository/data/gateway_config_expected.json [hunk 2]
```diff
@@ -882,13 +882,13 @@
...
-			"schema": "./middlewares/example/example_schema.json",
-			"importPath": "github.com/uber/zanzibar/examples/example-gateway/middlewares/example"
+			"importPath": "github.com/uber/zanzibar/examples/example-gateway/middlewares/example",
+			"schema": "./middlewares/example/example_schema.json"
 		},
 		"example_reader": {
 			"name": "example_reader",
-			"schema": "./middlewares/example_reader/example_reader_schema.json",
-			"importPath": "github.com/uber/zanzibar/examples/example-gateway/middlewares/example_reader"
+			"importPath": "github.com/uber/zanzibar/examples/example-gateway/middlewares/example_reader",
```
Comment (Matt-Esch) [inline backend/repository/data/gateway_config_expected.json]: I'm surprised that this is needed, the module system should provide import paths by convention (and it's part of the basic information provided by all module instances)

File: codegen/templates/endpoint.tmpl [hunk 3]
```diff
@@ -68,7 +68,7 @@ func New{{$handlerName}}(
 		{{ if len $middlewares | ne 0 -}}
 		zanzibar.NewStack([]zanzibar.MiddlewareHandle{
 			{{range $idx, $middleware := $middlewares -}}
-			{{$middleware.Name}}.NewMiddleWare(
+			deps.Middleware.{{$middleware.Name | pascal}}.NewMiddlewareHandle(
```
Comment (Matt-Esch) [inline codegen/templates/endpoint.tmpl]: the need to have pascal probably means we're going to run into convention errors at some point

File: codegen/gateway.go [hunk 4]
```diff
@@ -112,24 +112,64 @@ type ClientClassConfig struct {
...
-	Dependencies ClientDependencies     `json:"dependencies"`
+	Dependencies Dependencies           `json:"dependencies"`
 }
 
-// ClientDependencies lists all dependencies of a client.
-type ClientDependencies struct {
+// Dependencies lists all dependencies of a module
+type Dependencies struct {
 	Client []string `json:"client"`
+	//	Service []string `json:"service"`  // example extension
```
Comment (oliviazhang0809) [inline codegen/gateway.go]: Should it be better naming as `ClientDependencies`. Not sure about the usecase of service dependencies.

File: codegen/gateway.go [hunk 5]
```diff
@@ -315,54 +355,28 @@ func newClientSpec(
...
-	Name string `json:"name"`
-	// Go import path for the middleware.
-	Path string `json:"path,omitempty"`
+	Name string
 	// Middleware specific configuration options.
-	Options map[string]interface{} `json:"options,omitempty"`
+	Options map[string]interface{}
 	// Options pretty printed for template initialization
-	PrettyOptions map[string]string `json:"prettyOptions,omitempty"`
+	PrettyOptions map[string]string
```
Comment (cl1337) [inline codegen/gateway.go]: why required now?

File: backend/repository/data/gateway_config_expected.json (hunk 2 above)
Comment (jakelarkn) [reply backend/repository/data/gateway_config_expected.json]: holdover from the old middleware way-of-doing-things,  will take a stab at rm

File: codegen/module.go [hunk 6]
```diff
@@ -1098,7 +1098,7 @@ type ModuleInstance struct {
 	InstanceName string
 	// Config is a reference to the instance "config" key in the instances json
```
Comment (oliviazhang0809) [inline codegen/module.go]: this description needs to be updated

File: codegen/module.go [hunk 7]
```diff
@@ -1139,7 +1139,7 @@ type JSONClassConfig struct {
 	Name string `json:"name"`
 	// The configuration object for this class instance. This depends on the
```
Comment (oliviazhang0809) [inline codegen/module.go]: update description as well

File: codegen/module_system.go [hunk 8]
```diff
@@ -1054,6 +1076,80 @@ func (generator *GatewayServiceGenerator) Generate(
...
+	//meta := &MiddlewareMeta{
+	//	Instance: instance,
+	//	Go: instance.Config.(*MiddlewareConfigTmp),
+	//}
+	bytes, err := g.templates.ExecTemplate("middleware.tmpl", instance, g.packageHelper)
+	if err != nil {
+		return err
+	}
+
+	//targetMiddlewareFile := filepath.Join(
```
Comment (oliviazhang0809) [inline codegen/module_system.go]: remove commented out code or add TODO?

File: codegen/module_system.go [hunk 9]
```diff
@@ -1054,6 +1076,80 @@ func (generator *GatewayServiceGenerator) Generate(
...
+}
+
+// Generate returns the endpoint build result, which contains a file per
+// endpoint handler and a list of handler specs
+func (g *MiddlewareGenerator) Generate(
+	instance *ModuleInstance,
+) (*BuildResult, error) {
+	ret := map[string][]byte{}
+
+	//	clientSpecs := readClientDependencySpecs(instance)
```
Comment (cl1337) [inline codegen/module_system.go]: remove comment?

File: codegen/module_system.go [hunk 10]
```diff
@@ -1054,6 +1076,80 @@ func (generator *GatewayServiceGenerator) Generate(
...
+	}
+	return &BuildResult{
+		Files: ret,
+		Spec:  nil,
+	}, nil
+}
+
+func (g *MiddlewareGenerator) generateMiddlewareFile(instance *ModuleInstance, out map[string][]byte) error {
+
+	//meta := &MiddlewareMeta{
```
Comment (cl1337) [inline codegen/module_system.go]: is these comments needed?

File: runtime/middlewares_test.go [hunk 11]
```diff
@@ -36,8 +36,9 @@ import (
 
 // Ensures that a middleware stack can correctly return all of its handlers.
 func TestHandlers(t *testing.T) {
-	ex := example.NewMiddleWare(
+	ex := example.NewMiddleware(
 		nil, // *zanzibar.Gateway
+		nil,
```
Comment (oliviazhang0809) [inline runtime/middlewares_test.go]: can we add test with dep.client not null?

**Comment by oliviazhang0809** [comment]
nit: pls `make generate` before land to make copyright 2018. :)

File: test/lib/compare.go (hunk 1 above)
Comment (jakelarkn) [reply test/lib/compare.go]: nice

File: codegen/gateway.go (hunk 4 above)
Comment (jakelarkn) [reply codegen/gateway.go]: naming just mirrors the module system JSON config

service interfaces abstract (one or more clients + minimal business logic + feature toggles).  service dependencies are used somewhat extensively in RTAPI and are a part of the design of the module system in zanzibar.   they don't really make a lot of sense for the edge proxy use-cases   but imagine they will be important for fulfillment and other gateways as code base matures

File: codegen/gateway.go (hunk 5 above)
Comment (jakelarkn) [reply codegen/gateway.go]: I believe this object is no longer serialized with my changes in the repository package

File: codegen/module_system.go (hunk 10 above)
Comment (jakelarkn) [reply codegen/module_system.go]: had uncommited changes to remove them, doh!

**Comment by jakelarkn** [comment]
going to punt on the import changes for later PR.   hit a weird bug with identifying import path in codegen conflicting with how jenkins test run script utilizes a symlinked project directory,  resulting in a `make check-generate` that passes locally but fails on jenkins.  already lost a day on it ... this PR unblocks a ton of middleware work in edge.

**Pull request merged**